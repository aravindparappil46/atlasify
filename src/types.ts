import type {
  AtlasifyNotification,
  Category,
  Product,
  ReadState,
} from './utils/api/typesGitHub';
import type { AuthMethod, PlatformType } from './utils/auth/types';

declare const __brand: unique symbol;

type Brand<B> = { [__brand]: B };

export interface AuthState {
  accounts: Account[];
}

export type Branded<T, B> = T & Brand<B>;

export type Username = Branded<string, 'Username'>;

export type Token = Branded<string, 'Token'>;

export type Organization = Branded<string, 'Organization'>;

export type Link = Branded<string, 'WebUrl'>;

export type Status = 'loading' | 'success' | 'error';

export interface Account {
  method: AuthMethod;
  platform: PlatformType;
  token: Token;
  user: AtlasifyUser;
}

export type SettingsValue =
  | boolean
  | number
  | GroupBy
  | OpenPreference
  | Category[]
  | ReadState[]
  | Product[]
  | Theme;

export type SettingsState = AppearanceSettingsState &
  NotificationSettingsState &
  SystemSettingsState &
  FilterSettingsState;

interface AppearanceSettingsState {
  theme: Theme;
  zoomPercentage: number;
  showAccountHeader: boolean;
}

interface NotificationSettingsState {
  groupBy: GroupBy;
  delayNotificationState: boolean;
}

interface SystemSettingsState {
  openLinks: OpenPreference;
  keyboardShortcut: boolean;
  showNotificationsCountInTray: boolean;
  showSystemNotifications: boolean;
  useAlternateIdleIcon: boolean;
  playSound: boolean;
  openAtStartup: boolean;
}

interface FilterSettingsState {
  filterCategories: Category[];
  filterReadStates: ReadState[];
  filterProducts: Product[];
}

export interface AtlasifyState {
  auth?: AuthState;
  settings?: SettingsState;
}

export enum Theme {
  SYSTEM = 'SYSTEM',
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}

export enum OpenPreference {
  FOREGROUND = 'FOREGROUND',
  BACKGROUND = 'FOREGROUND',
}

export enum GroupBy {
  PRODUCT = 'PRODUCT',
  REPOSITORY = 'REPOSITORY',
  DATE = 'DATE',
}

export type RadioGroupItem = {
  label: string;
  value: string;
};

export interface AccountNotifications {
  account: Account;
  notifications: AtlasifyNotification[];
  error: AtlasifyError | null;
}

export interface AtlasifyUser {
  login: string;
  name: string | null;
  avatar: Link | null;
  id: string;
}

export interface AtlasifyError {
  title: string;
  descriptions: string[];
  emojis: string[];
}

export type ErrorType =
  | 'BAD_CREDENTIALS'
  | 'MISSING_SCOPES'
  | 'NETWORK'
  | 'RATE_LIMITED'
  | 'UNKNOWN';

export interface FormattedReason {
  title: string;
  description: string;
}

// TODO - Remove this
export enum IconColor {
  GRAY = 'text-gray-500 dark:text-gray-300',
  GREEN = 'text-green-500',
  PURPLE = 'text-purple-500',
  RED = 'text-red-500',
  YELLOW = 'text-yellow-500 dark:text-yellow-300',
  WHITE = 'text-white',
}

export enum Opacity {
  READ = 'opacity-50',
  LOW = 'opacity-70',
  MEDIUM = 'opacity-80',
  HIGH = 'opacity-90',
}

export enum Size {
  XSMALL = 12,
  SMALL = 14,
  MEDIUM = 16,
  LARGE = 18,
  XLARGE = 20,
}
