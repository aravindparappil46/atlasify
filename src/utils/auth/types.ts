import type { Token, Username } from '../../types';

export type AuthMethod = 'API Token';

export type PlatformType = 'Atlassian Cloud';

export interface LoginAPITokenOptions {
  username: Username;
  token: Token;
}
