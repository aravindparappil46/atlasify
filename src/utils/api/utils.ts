import { Constants } from '../constants';

export function getAPIUrl(): URL {
  return new URL(Constants.ATLASSIAN_API);
}
