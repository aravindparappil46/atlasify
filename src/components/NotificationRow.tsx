import { type FC, useCallback, useContext, useState } from 'react';

import Avatar from '@atlaskit/avatar';
import { IconButton } from '@atlaskit/button/new';
import HipchatMediaAttachmentCountIcon from '@atlaskit/icon/glyph/hipchat/media-attachment-count';
import Tooltip from '@atlaskit/tooltip';

import { AppContext } from '../context/App';
import { type AtlasifyNotification, Opacity } from '../types';
import { cn } from '../utils/cn';
import {
  formatNotificationUpdatedAt,
  formatProperCase,
} from '../utils/helpers';
import { openNotification } from '../utils/links';
import { HoverGroup } from './HoverGroup';

interface INotificationRow {
  notification: AtlasifyNotification;
  isAnimated?: boolean;
  isRead?: boolean;
}

export const NotificationRow: FC<INotificationRow> = ({
  notification,
  isAnimated = false,
  isRead = false,
}: INotificationRow) => {
  const { settings, markNotificationRead } = useContext(AppContext);
  const [animateExit, setAnimateExit] = useState(false);
  const [showAsRead, setShowAsRead] = useState(false);

  const handleNotification = useCallback(() => {
    setAnimateExit(!settings.delayNotificationState);

    if (settings.markAsReadOnOpen) {
      setShowAsRead(settings.delayNotificationState);
      markNotificationRead(notification);
    }

    openNotification(notification);
  }, [notification, markNotificationRead, settings]);

  const updatedAt = formatNotificationUpdatedAt(notification);

  return (
    <div
      id={notification.id}
      className={cn(
        'group flex border-b border-gray-100 bg-white px-3 py-2 hover:bg-gray-100 dark:border-gray-darker dark:bg-gray-dark dark:text-white dark:hover:bg-gray-darker',
        (isAnimated || animateExit) &&
          'translate-x-full opacity-0 transition duration-[350ms] ease-in-out',
        (isRead || showAsRead) && Opacity.READ,
      )}
    >
      <div className="mr-3 flex items-center justify-center">
        <Tooltip content={notification.actor.displayName}>
          <Avatar
            name={notification.actor.displayName}
            src={notification.actor.avatarURL}
            size="small"
          />
        </Tooltip>
      </div>

      <div
        className="mr-3 flex flex-1 cursor-pointer text-wrap"
        onClick={() => handleNotification()}
      >
        <div className="flex flex-col gap-1">
          <div
            className="text-sm"
            role="main"
            title={notification.title.trim()}
          >
            {notification.title}
            <span
              className="pl-2 text-xs text-gray-700"
              title={`Updated ${updatedAt}`}
            >
              {updatedAt}
            </span>
          </div>

          <div
            className={cn(
              'flex flex-1 items-center gap-1 text-xs',
              Opacity.MEDIUM,
            )}
          >
            <Avatar
              name={notification.entity.title}
              src={notification.entity.iconUrl}
              size="xsmall"
              appearance="square"
            />
            {notification.entity.title}
          </div>

          <div
            className={cn(
              'flex flex-1 items-center pl-0.5 gap-1.5 text-xs',
              Opacity.MEDIUM,
            )}
          >
            <notification.product.icon size="xsmall" appearance="brand" />
            {notification.path?.title ??
              formatProperCase(notification.product.name)}
          </div>
        </div>
      </div>

      {!animateExit && (
        <HoverGroup>
          {notification.unread && (
            <IconButton
              icon={(iconProps) => (
                <HipchatMediaAttachmentCountIcon {...iconProps} size="small" />
              )}
              label="Mark as read"
              isTooltipDisabled={false}
              shape="circle"
              spacing="compact"
              appearance="subtle"
              onClick={() => {
                setAnimateExit(!settings.delayNotificationState);
                setShowAsRead(settings.delayNotificationState);
                markNotificationRead(notification);
              }}
            />
          )}
        </HoverGroup>
      )}
    </div>
  );
};
