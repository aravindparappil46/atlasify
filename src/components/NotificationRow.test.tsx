import { fireEvent, render, screen } from '@testing-library/react';
import {
  mockAtlassianCloudAccount,
  mockAuth,
  mockSettings,
} from '../__mocks__/state-mocks';
import { AppContext } from '../context/App';
import { GroupBy } from '../types';
import { mockSingleNotification } from '../utils/api/__mocks__/response-mocks';
import * as comms from '../utils/comms';
import * as links from '../utils/links';
import { NotificationRow } from './NotificationRow';

describe('components/NotificationRow.tsx', () => {
  jest.spyOn(links, 'openNotification');
  jest.spyOn(comms, 'openExternalLink').mockImplementation();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render itself & its children - group by date', async () => {
    jest
      .spyOn(global.Date, 'now')
      .mockImplementation(() => new Date('2024').valueOf());

    const props = {
      notification: mockSingleNotification,
      account: mockAtlassianCloudAccount,
    };

    const tree = render(
      <AppContext.Provider
        value={{ settings: { ...mockSettings, groupBy: GroupBy.DATE } }}
      >
        <NotificationRow {...props} />
      </AppContext.Provider>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render itself & its children - group by repositories', async () => {
    jest
      .spyOn(global.Date, 'now')
      .mockImplementation(() => new Date('2024').valueOf());

    const props = {
      notification: mockSingleNotification,
      account: mockAtlassianCloudAccount,
    };

    const tree = render(
      <AppContext.Provider
        value={{ settings: { ...mockSettings, groupBy: GroupBy.DATE } }}
      >
        <NotificationRow {...props} />
      </AppContext.Provider>,
    );
    expect(tree).toMatchSnapshot();
  });

  describe('notification interactions', () => {
    it('should open a notification in the browser - click', () => {
      const markNotificationRead = jest.fn();

      const props = {
        notification: mockSingleNotification,
        account: mockAtlassianCloudAccount,
      };

      render(
        <AppContext.Provider
          value={{
            settings: mockSettings,
            markNotificationRead,
            auth: mockAuth,
          }}
        >
          <NotificationRow {...props} />
        </AppContext.Provider>,
      );

      fireEvent.click(screen.getByRole('main'));
      expect(links.openNotification).toHaveBeenCalledTimes(1);
      expect(markNotificationRead).toHaveBeenCalledTimes(1);
    });

    it('should open a notification in the browser - delay notification setting enabled', () => {
      const markNotificationRead = jest.fn();

      const props = {
        notification: mockSingleNotification,
        account: mockAtlassianCloudAccount,
      };

      render(
        <AppContext.Provider
          value={{
            settings: {
              ...mockSettings,
              delayNotificationState: true,
            },
            markNotificationRead,
            auth: mockAuth,
          }}
        >
          <NotificationRow {...props} />
        </AppContext.Provider>,
      );

      fireEvent.click(screen.getByRole('main'));
      expect(links.openNotification).toHaveBeenCalledTimes(1);
      expect(markNotificationRead).toHaveBeenCalledTimes(1);
    });

    it('should open a notification in the browser - key down', () => {
      const markNotificationRead = jest.fn();

      const props = {
        notification: mockSingleNotification,
        account: mockAtlassianCloudAccount,
      };

      render(
        <AppContext.Provider
          value={{
            settings: mockSettings,
            markNotificationRead,
            auth: mockAuth,
          }}
        >
          <NotificationRow {...props} />
        </AppContext.Provider>,
      );

      fireEvent.click(screen.getByRole('main'));
      expect(links.openNotification).toHaveBeenCalledTimes(1);
      expect(markNotificationRead).toHaveBeenCalledTimes(1);
    });

    it('should mark a notification as read', () => {
      const markNotificationRead = jest.fn();

      const props = {
        notification: mockSingleNotification,
        account: mockAtlassianCloudAccount,
      };

      render(
        <AppContext.Provider
          value={{
            settings: mockSettings,
            markNotificationRead,
          }}
        >
          <NotificationRow {...props} />
        </AppContext.Provider>,
      );

      fireEvent.click(screen.getByTitle('Mark as read'));
      expect(markNotificationRead).toHaveBeenCalledTimes(1);
    });
  });
});
