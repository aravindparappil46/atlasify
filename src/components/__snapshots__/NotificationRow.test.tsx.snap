// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/NotificationRow.tsx should render itself & its children - group by date 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="group flex border-b border-gray-100 bg-white px-3 py-2 hover:bg-gray-100 dark:border-gray-darker dark:bg-gray-dark dark:text-white dark:hover:bg-gray-darker"
        id="138661096"
      >
        <div
          class="mr-3 flex items-center justify-center"
        >
          <div
            role="presentation"
          >
            <div
              aria-labelledby="2"
              role="img"
              style="display: inline-block; position: relative; outline: 0;"
            >
              <span
                class="css-1hqd2wf"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  class="css-13ep12v"
                  data-vc="avatar-image"
                  src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
                  style="border-radius: 50%;"
                />
              </span>
              <span
                hidden=""
                id="2"
              >
                gitify-app
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex-1 truncate cursor-pointer"
        >
          <div
            class="mb-1 flex items-center gap-1 text-xs font-medium opacity-80"
          >
            <span>
              <img
                alt="atlasify-app/notifications-test's avatar"
                class="rounded-full object-cover size-4"
                src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
              />
            </span>
            <span
              class="cursor-pointer truncate opacity-90"
              title="atlasify-app/notifications-test"
            >
              atlasify-app/notifications-test
            </span>
          </div>
          <div
            class="flex gap-1 items-center mb-1 truncate text-sm"
            role="main"
            title="I am a robot and this is a test!"
          >
            <span
              class="truncate"
            >
              I am a robot and this is a test!
            </span>
            <span
              class="text-xs"
              title="gitify-app updated 7 years ago"
            >
              7 years ago
            </span>
          </div>
          <div>
            <div
              class="flex flex-wrap items-center gap-1 text-xs opacity-80"
            >
              <div
                aria-labelledby="3"
                role="img"
                style="display: inline-block; position: relative; outline: 0;"
              >
                <span
                  class="css-122fgfa"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    class="css-13ep12v"
                    data-vc="avatar-image"
                    src="https://github.atlasify.io/favicon.ico"
                    style="border-radius: 2px;"
                  />
                </span>
                <span
                  hidden=""
                  id="3"
                >
                  Repository
                </span>
              </div>
              <div>
                Repository
              </div>
            </div>
            <div
              class="flex flex-wrap items-center gap-1 text-xs opacity-80"
            >
              <span
                aria-label="Bitbucket"
                class="css-5vjqu0"
                role="img"
                style="--logo-color: inherit; --logo-fill: currentColor;"
              >
                <svg
                  aria-hidden="true"
                  fill="none"
                  focusable="false"
                  height="32"
                  viewBox="0 0 32 32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
      
                  <path
                    d="m25.586 15.819-1.618 9.885c-.106.598-.528.95-1.126.95H9.122c-.598 0-1.02-.352-1.126-.95L5.146 8.08c-.105-.598.212-.985.775-.985h20.123c.562 0 .88.387.773.985l-.773 4.644c-.106.668-.493.95-1.126.95H12.816c-.176 0-.282.105-.246.316l.95 5.84c.035.14.14.246.281.246h4.362c.141 0 .246-.105.282-.246l.668-4.222c.07-.527.422-.738.915-.738h4.75c.703 0 .914.351.808.95"
                    fill="#1868DB"
                  />
                  
    
                </svg>
              </span>
              Repository
            </div>
          </div>
        </div>
        <div
          class="flex items-center justify-center gap-2 opacity-0 transition-opacity group-hover:opacity-80"
        >
          <button
            class="h-full hover:text-green-500 focus:outline-none"
            title="Mark as Read"
            type="button"
          >
            <svg
              aria-label="Mark as Read"
              class="octicon octicon-read"
              fill="currentColor"
              focusable="false"
              height="14"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="14"
            >
              <path
                d="M7.115.65a1.752 1.752 0 0 1 1.77 0l6.25 3.663c.536.314.865.889.865 1.51v6.427A1.75 1.75 0 0 1 14.25 14H1.75A1.75 1.75 0 0 1 0 12.25V5.823c0-.621.33-1.196.865-1.51Zm1.011 1.293a.252.252 0 0 0-.252 0l-5.72 3.353L6.468 7.76a2.748 2.748 0 0 1 3.066 0l4.312-2.464-5.719-3.353ZM13.15 12.5 8.772 9.06a1.25 1.25 0 0 0-1.544 0L2.85 12.5Zm1.35-5.85-3.687 2.106 3.687 2.897ZM5.187 8.756 1.5 6.65v5.003Z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="group flex border-b border-gray-100 bg-white px-3 py-2 hover:bg-gray-100 dark:border-gray-darker dark:bg-gray-dark dark:text-white dark:hover:bg-gray-darker"
      id="138661096"
    >
      <div
        class="mr-3 flex items-center justify-center"
      >
        <div
          role="presentation"
        >
          <div
            aria-labelledby="2"
            role="img"
            style="display: inline-block; position: relative; outline: 0;"
          >
            <span
              class="css-1hqd2wf"
            >
              <img
                alt=""
                aria-hidden="true"
                class="css-13ep12v"
                data-vc="avatar-image"
                src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
                style="border-radius: 50%;"
              />
            </span>
            <span
              hidden=""
              id="2"
            >
              gitify-app
            </span>
          </div>
        </div>
      </div>
      <div
        class="flex-1 truncate cursor-pointer"
      >
        <div
          class="mb-1 flex items-center gap-1 text-xs font-medium opacity-80"
        >
          <span>
            <img
              alt="atlasify-app/notifications-test's avatar"
              class="rounded-full object-cover size-4"
              src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
            />
          </span>
          <span
            class="cursor-pointer truncate opacity-90"
            title="atlasify-app/notifications-test"
          >
            atlasify-app/notifications-test
          </span>
        </div>
        <div
          class="flex gap-1 items-center mb-1 truncate text-sm"
          role="main"
          title="I am a robot and this is a test!"
        >
          <span
            class="truncate"
          >
            I am a robot and this is a test!
          </span>
          <span
            class="text-xs"
            title="gitify-app updated 7 years ago"
          >
            7 years ago
          </span>
        </div>
        <div>
          <div
            class="flex flex-wrap items-center gap-1 text-xs opacity-80"
          >
            <div
              aria-labelledby="3"
              role="img"
              style="display: inline-block; position: relative; outline: 0;"
            >
              <span
                class="css-122fgfa"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  class="css-13ep12v"
                  data-vc="avatar-image"
                  src="https://github.atlasify.io/favicon.ico"
                  style="border-radius: 2px;"
                />
              </span>
              <span
                hidden=""
                id="3"
              >
                Repository
              </span>
            </div>
            <div>
              Repository
            </div>
          </div>
          <div
            class="flex flex-wrap items-center gap-1 text-xs opacity-80"
          >
            <span
              aria-label="Bitbucket"
              class="css-5vjqu0"
              role="img"
              style="--logo-color: inherit; --logo-fill: currentColor;"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="32"
                viewBox="0 0 32 32"
                xmlns="http://www.w3.org/2000/svg"
              >
                
      
                <path
                  d="m25.586 15.819-1.618 9.885c-.106.598-.528.95-1.126.95H9.122c-.598 0-1.02-.352-1.126-.95L5.146 8.08c-.105-.598.212-.985.775-.985h20.123c.562 0 .88.387.773.985l-.773 4.644c-.106.668-.493.95-1.126.95H12.816c-.176 0-.282.105-.246.316l.95 5.84c.035.14.14.246.281.246h4.362c.141 0 .246-.105.282-.246l.668-4.222c.07-.527.422-.738.915-.738h4.75c.703 0 .914.351.808.95"
                  fill="#1868DB"
                />
                
    
              </svg>
            </span>
            Repository
          </div>
        </div>
      </div>
      <div
        class="flex items-center justify-center gap-2 opacity-0 transition-opacity group-hover:opacity-80"
      >
        <button
          class="h-full hover:text-green-500 focus:outline-none"
          title="Mark as Read"
          type="button"
        >
          <svg
            aria-label="Mark as Read"
            class="octicon octicon-read"
            fill="currentColor"
            focusable="false"
            height="14"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="14"
          >
            <path
              d="M7.115.65a1.752 1.752 0 0 1 1.77 0l6.25 3.663c.536.314.865.889.865 1.51v6.427A1.75 1.75 0 0 1 14.25 14H1.75A1.75 1.75 0 0 1 0 12.25V5.823c0-.621.33-1.196.865-1.51Zm1.011 1.293a.252.252 0 0 0-.252 0l-5.72 3.353L6.468 7.76a2.748 2.748 0 0 1 3.066 0l4.312-2.464-5.719-3.353ZM13.15 12.5 8.772 9.06a1.25 1.25 0 0 0-1.544 0L2.85 12.5Zm1.35-5.85-3.687 2.106 3.687 2.897ZM5.187 8.756 1.5 6.65v5.003Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/NotificationRow.tsx should render itself & its children - group by repositories 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="group flex border-b border-gray-100 bg-white px-3 py-2 hover:bg-gray-100 dark:border-gray-darker dark:bg-gray-dark dark:text-white dark:hover:bg-gray-darker"
        id="138661096"
      >
        <div
          class="mr-3 flex items-center justify-center"
        >
          <div
            role="presentation"
          >
            <div
              aria-labelledby="5"
              role="img"
              style="display: inline-block; position: relative; outline: 0;"
            >
              <span
                class="css-1hqd2wf"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  class="css-13ep12v"
                  data-vc="avatar-image"
                  src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
                  style="border-radius: 50%;"
                />
              </span>
              <span
                hidden=""
                id="5"
              >
                gitify-app
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex-1 truncate cursor-pointer"
        >
          <div
            class="flex gap-1 items-center mb-1 truncate text-sm"
            role="main"
            title="I am a robot and this is a test!"
          >
            <span
              class="truncate"
            >
              I am a robot and this is a test!
            </span>
            <span
              class="text-xs"
              title="gitify-app updated 7 years ago"
            >
              7 years ago
            </span>
          </div>
          <div>
            <div
              class="flex flex-wrap items-center gap-1 text-xs opacity-80"
            >
              <div
                aria-labelledby="6"
                role="img"
                style="display: inline-block; position: relative; outline: 0;"
              >
                <span
                  class="css-122fgfa"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    class="css-13ep12v"
                    data-vc="avatar-image"
                    src="https://github.atlasify.io/favicon.ico"
                    style="border-radius: 2px;"
                  />
                </span>
                <span
                  hidden=""
                  id="6"
                >
                  Repository
                </span>
              </div>
              <div>
                Repository
              </div>
            </div>
            <div
              class="flex flex-wrap items-center gap-1 text-xs opacity-80"
            >
              <span
                aria-label="Bitbucket"
                class="css-5vjqu0"
                role="img"
                style="--logo-color: inherit; --logo-fill: currentColor;"
              >
                <svg
                  aria-hidden="true"
                  fill="none"
                  focusable="false"
                  height="32"
                  viewBox="0 0 32 32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
      
                  <path
                    d="m25.586 15.819-1.618 9.885c-.106.598-.528.95-1.126.95H9.122c-.598 0-1.02-.352-1.126-.95L5.146 8.08c-.105-.598.212-.985.775-.985h20.123c.562 0 .88.387.773.985l-.773 4.644c-.106.668-.493.95-1.126.95H12.816c-.176 0-.282.105-.246.316l.95 5.84c.035.14.14.246.281.246h4.362c.141 0 .246-.105.282-.246l.668-4.222c.07-.527.422-.738.915-.738h4.75c.703 0 .914.351.808.95"
                    fill="#1868DB"
                  />
                  
    
                </svg>
              </span>
              Repository
            </div>
          </div>
        </div>
        <div
          class="flex items-center justify-center gap-2 opacity-0 transition-opacity group-hover:opacity-80"
        >
          <button
            class="h-full hover:text-green-500 focus:outline-none"
            title="Mark as Read"
            type="button"
          >
            <svg
              aria-label="Mark as Read"
              class="octicon octicon-read"
              fill="currentColor"
              focusable="false"
              height="14"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="14"
            >
              <path
                d="M7.115.65a1.752 1.752 0 0 1 1.77 0l6.25 3.663c.536.314.865.889.865 1.51v6.427A1.75 1.75 0 0 1 14.25 14H1.75A1.75 1.75 0 0 1 0 12.25V5.823c0-.621.33-1.196.865-1.51Zm1.011 1.293a.252.252 0 0 0-.252 0l-5.72 3.353L6.468 7.76a2.748 2.748 0 0 1 3.066 0l4.312-2.464-5.719-3.353ZM13.15 12.5 8.772 9.06a1.25 1.25 0 0 0-1.544 0L2.85 12.5Zm1.35-5.85-3.687 2.106 3.687 2.897ZM5.187 8.756 1.5 6.65v5.003Z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="group flex border-b border-gray-100 bg-white px-3 py-2 hover:bg-gray-100 dark:border-gray-darker dark:bg-gray-dark dark:text-white dark:hover:bg-gray-darker"
      id="138661096"
    >
      <div
        class="mr-3 flex items-center justify-center"
      >
        <div
          role="presentation"
        >
          <div
            aria-labelledby="5"
            role="img"
            style="display: inline-block; position: relative; outline: 0;"
          >
            <span
              class="css-1hqd2wf"
            >
              <img
                alt=""
                aria-hidden="true"
                class="css-13ep12v"
                data-vc="avatar-image"
                src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
                style="border-radius: 50%;"
              />
            </span>
            <span
              hidden=""
              id="5"
            >
              gitify-app
            </span>
          </div>
        </div>
      </div>
      <div
        class="flex-1 truncate cursor-pointer"
      >
        <div
          class="flex gap-1 items-center mb-1 truncate text-sm"
          role="main"
          title="I am a robot and this is a test!"
        >
          <span
            class="truncate"
          >
            I am a robot and this is a test!
          </span>
          <span
            class="text-xs"
            title="gitify-app updated 7 years ago"
          >
            7 years ago
          </span>
        </div>
        <div>
          <div
            class="flex flex-wrap items-center gap-1 text-xs opacity-80"
          >
            <div
              aria-labelledby="6"
              role="img"
              style="display: inline-block; position: relative; outline: 0;"
            >
              <span
                class="css-122fgfa"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  class="css-13ep12v"
                  data-vc="avatar-image"
                  src="https://github.atlasify.io/favicon.ico"
                  style="border-radius: 2px;"
                />
              </span>
              <span
                hidden=""
                id="6"
              >
                Repository
              </span>
            </div>
            <div>
              Repository
            </div>
          </div>
          <div
            class="flex flex-wrap items-center gap-1 text-xs opacity-80"
          >
            <span
              aria-label="Bitbucket"
              class="css-5vjqu0"
              role="img"
              style="--logo-color: inherit; --logo-fill: currentColor;"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="32"
                viewBox="0 0 32 32"
                xmlns="http://www.w3.org/2000/svg"
              >
                
      
                <path
                  d="m25.586 15.819-1.618 9.885c-.106.598-.528.95-1.126.95H9.122c-.598 0-1.02-.352-1.126-.95L5.146 8.08c-.105-.598.212-.985.775-.985h20.123c.562 0 .88.387.773.985l-.773 4.644c-.106.668-.493.95-1.126.95H12.816c-.176 0-.282.105-.246.316l.95 5.84c.035.14.14.246.281.246h4.362c.141 0 .246-.105.282-.246l.668-4.222c.07-.527.422-.738.915-.738h4.75c.703 0 .914.351.808.95"
                  fill="#1868DB"
                />
                
    
              </svg>
            </span>
            Repository
          </div>
        </div>
      </div>
      <div
        class="flex items-center justify-center gap-2 opacity-0 transition-opacity group-hover:opacity-80"
      >
        <button
          class="h-full hover:text-green-500 focus:outline-none"
          title="Mark as Read"
          type="button"
        >
          <svg
            aria-label="Mark as Read"
            class="octicon octicon-read"
            fill="currentColor"
            focusable="false"
            height="14"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="14"
          >
            <path
              d="M7.115.65a1.752 1.752 0 0 1 1.77 0l6.25 3.663c.536.314.865.889.865 1.51v6.427A1.75 1.75 0 0 1 14.25 14H1.75A1.75 1.75 0 0 1 0 12.25V5.823c0-.621.33-1.196.865-1.51Zm1.011 1.293a.252.252 0 0 0-.252 0l-5.72 3.353L6.468 7.76a2.748 2.748 0 0 1 3.066 0l4.312-2.464-5.719-3.353ZM13.15 12.5 8.772 9.06a1.25 1.25 0 0 0-1.544 0L2.85 12.5Zm1.35-5.85-3.687 2.106 3.687 2.897ZM5.187 8.756 1.5 6.65v5.003Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
